From 614ffe685b5d891db5b67b025f809b31aa3b966d Mon Sep 17 00:00:00 2001
From: Jackal Guo <jackalguo@google.com>
Date: Fri, 24 Dec 2021 11:47:47 +0800
Subject: [PATCH] Fix the inconsistency of protection level

The value may be inconsistent between BasePermission.perm.protection-
Level and BasePermission.getProtectionLevel() within the same object.
Update the perm to fix the inconsistency.

Bug: 209607944
Test: manually using the PoC on the buganizer to ensure the symptom
      no longer exists.
Change-Id: I19d6135f98bee9392d85e56478c42e06cfea8ba5
(cherry picked from commit b5efdf729385cc54f225496d3ba20f1cb5b68250)
Merged-In:I19d6135f98bee9392d85e56478c42e06cfea8ba5
---
 .../java/com/android/server/pm/permission/BasePermission.java   | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/services/core/java/com/android/server/pm/permission/BasePermission.java b/services/core/java/com/android/server/pm/permission/BasePermission.java
index 5e04171a3bca..354f5b9a8bdf 100644
--- a/services/core/java/com/android/server/pm/permission/BasePermission.java
+++ b/services/core/java/com/android/server/pm/permission/BasePermission.java
@@ -409,6 +409,8 @@ public final class BasePermission {
         }
         if (bp.perm != null && Objects.equals(bp.perm.getPackageName(), p.getPackageName())
                 && Objects.equals(bp.perm.getName(), p.getName())) {
+            bp.perm.setFlags(bp.perm.getFlags() & ~PermissionInfo.FLAG_INSTALLED);
+            bp.perm = p.setFlags(p.getFlags() | PermissionInfo.FLAG_INSTALLED);
             bp.protectionLevel = p.getProtectionLevel();
         }
         if (bp.isRuntime() && (ownerChanged || wasNonRuntime)) {
-- 
2.17.1

